internal/cephfs/groupcontrollerserver.go

```diff
@@ -61,3 +72,563 @@ func (cs *ControllerServer) validateVolumeGroupSnapshotRequest(
+			Secrets:        secrets,
+			Parameters: map[string]string{
+				"clusterID": clusterID,
+			},
+		}
+
+		resp, err = cs.createSnapshotAndAddMapping(ctx, req, vgo, vgs)
+		if err != nil {
+			// Handle cleanup
+			log.ErrorLog(ctx, "failed to create snapshot: %v", err)
+
+			return nil, err
+		}
+		responses = append(responses, *resp)
+	}
+
+	if !freeze {
+		err = releaseSubvolumeQuiesce(ctx, vgo.RequestName, fsMap)
+		if err != nil {
+			log.ErrorLog(ctx, "failed to freeze subvolume: %v", err)
+
+			return nil, err
+		}
+	}
+
+	return &responses, nil
+}
+
+// releaseSubvolumeQuiesce releases the quiesce of the subvolumes and subvolume
+// groups in the filesystems for the volumeID's present in the
+// CreateVolumeGroupSnapshotRequest.
+func releaseSubvolumeQuiesce(ctx context.Context,
+	requestName string,
+	fsMap map[string]core.SubVolumeQuiesceClient,
+) error {
+	releasing := false
+	var err error
+	var data *admin.QuiesceInfo
+	for _, fm := range fsMap {
+		// UnFreeze the filesystems, subvolumes and subvolume groups
+		data, err = fm.ReleaseSubvolumeQuiesce(ctx, requestName)
+		if err != nil && !errors.Is(err, rados.ErrPermissionDenied) {
+			log.ErrorLog(ctx, "failed to release subvolume quiesce: %v", err)
+
+			return err
+		}
+
+		state := core.GetQuiesceState(data)
+		if errors.Is(err, rados.ErrPermissionDenied) && state != core.Released {
+			releasing = true
+
+			continue
+		}
+	}
+
+	if releasing {
+		return cerrors.ErrInProgress
+	}
+
+	return nil
+}
+
+func subvolumeQuiesceWithExpireTimeout(ctx context.Context,
+	requestName string,
+	fsMap map[string]core.SubVolumeQuiesceClient,
+) error {
+	var err error
+
+	var data *admin.QuiesceInfo
+	inProgress := false
+	for _, fm := range fsMap {
+		// reinitialize the expiry timer for the quiesce
+		data, err = fm.SubvolumeQuiesceWithExpireTimeout(ctx, requestName)
+		if err != nil && !errors.Is(err, rados.ErrPermissionDenied) {
+			log.ErrorLog(ctx, "failed to quiesce subvolume with timeout: %v", err)
+
+			return err
+		}
+		state := core.GetQuiesceState(data)
+		if errors.Is(err, rados.ErrPermissionDenied) || state == core.Quiescing {
+			inProgress = true
+
+			continue
+		}
+
+		if state != core.Quiesced {
+			return fmt.Errorf("quiesce operation is in %s state", state)
+		}
+	}
+
+	if inProgress {
+		return cerrors.ErrInProgress
+	}
+
+	return nil
+}
+
+func (cs *ControllerServer) createSnapshotAndAddMapping(
+	ctx context.Context,
+	req *csi.CreateSnapshotRequest,
+	vgo *store.VolumeGroupOptions,
+	vgs *store.VolumeGroupSnapshotIdentifier,
+) (*csi.CreateSnapshotResponse, error) {
+	// Create the snapshot
+	resp, err := cs.CreateSnapshot(ctx, req)
+	if err != nil {
+		// Handle cleanup
+		log.ErrorLog(ctx, "failed to create snapshot: %v", err)
+
+		return nil, err
+	}
+	// Add the snapshot to the volume group journal
+	err = store.VolumeGroupJournal.AddVolumeSnapshotMapping(ctx,
+		vgo.MetadataPool,
+		vgs.ReservedID,
+		req.SourceVolumeId,
```

Comment:
all `req` usage should ideally use `.Get..()` functions for improved error handling/parsing in the gRPC layer.